File: perl/top5.pl
#!/usr/bin/perl

use strict;
use warnings;


sub main() {
    my %count;
    my @sorted;

    while(<>) {
        chomp;
        if( exists($count{$_}) ) {
            $count{$_}++;
        } else {
            $count{$_} = 1;
        }
    }

    @sorted = sort { $count{$b} <=> $count{$a} } keys(%count);

    for(my $i=0; $i<5; $i++) {
        my $line = sprintf("%4s %s\n", $sorted[$i], $count{$sorted[$i]});
        print $line;
    }
}

main unless caller;

File: python/top5.py
#!/usr/bin/env python

from collections import defaultdict
import sys

def readnumbers(filename):
    with open(filename) as input:
        for line in input:
            yield line.strip()

def tally_counts(filename):
    counter = defaultdict(int)
    for number in readnumbers(filename):
        counter[number] += 1
    for key, count in counter.items():
        yield count, key

def top5(filename):
    counts = tally_counts(filename)
    most_common = sorted(tally_counts(filename), reverse=True)
    return most_common[:5]

def main():
    for count, key in top5(sys.argv[1]):
        print "%4s %s" % (key, count)

if __name__ == '__main__':
    main()
